### PR TITLE
Add ability to handle DescriptorUpdateTemplate extension to unique objects layer

### PR DESCRIPTION
**Pull request by mark-lunarg**
The UniqueObjects utility layer needed to unwrap vulkan object handles embedded in a buffer parameter of `vkUpdateDescriptorSetWithTemplateKHR()`.  This patch series parses that buffer, replacing the uniquified handles with the actual ones before passing the call to the hardware.  This takes care of part of issue #1529.

### CODE REVIEW

File: layers/unique_objects.cpp [hunk 1]
```diff
@@ -647,25 +649,93 @@ VKAPI_ATTR void VKAPI_CALL DestroyDescriptorUpdateTemplateKHR(VkDevice device,
...
+                auto image_entry = reinterpret_cast<VkDescriptorImageInfo *>(entry);
+                size = std::max(size, (char *)offset + sizeof(VkDescriptorImageInfo));
+
+                auto sampler_handle = dev_data->unique_id_mapping[reinterpret_cast<uint64_t &>(image_entry->sampler)];
+                patch_list.emplace_back(std::make_pair(offset, sampler_handle));
+                offset += sizeof(VkSampler);
+                auto image_view_handle = dev_data->unique_id_mapping[reinterpret_cast<uint64_t &>(image_entry->imageView)];
+                patch_list.emplace_back(std::make_pair(offset, image_view_handle));
+                offset += sizeof(VkImageView);
+                patch_list.emplace_back(std::make_pair(offset, image_entry->imageLayout));
```
Comment (chrisforbes) [inline layers/unique_objects.cpp]: This emits a 64 bits fixup to a 32 bits value. You get away with it in a LE 64 bit build as you zero extend imageLayout and then scribble in the struct padding, but this is going to go badly for 32 bit builds, if this happens to be right at the end of the buffer.

[or on BE, if anyone decides to run the layers there-- but they probably also have other problems]

File: layers/unique_objects.cpp [hunk 2]
```diff
@@ -647,25 +649,93 @@ VKAPI_ATTR void VKAPI_CALL DestroyDescriptorUpdateTemplateKHR(VkDevice device,
...
+    if (template_map_entry == dev_data->desc_template_map.end()) {
+        assert(0);
+    }
+    auto const &create_info = template_map_entry->second->create_info;
+    char *size = 0;
+    std::vector<std::pair<char *, uint64_t>> patch_list;
+
+    for (uint32_t i = 0; i < create_info->descriptorUpdateEntryCount; i++) {
+        for (uint32_t j = 0; j < create_info->pDescriptorUpdateEntries[i].descriptorCount; j++) {
+            char *offset =
```
Comment (chrisforbes) [inline layers/unique_objects.cpp]: This is hard to follow -- you have offsets and sizes that are actually ptrs?

File: layers/unique_objects.h [hunk 3]
```diff
@@ -33,6 +33,14 @@ namespace unique_objects {
 // All increments must be guarded by global_lock
 static uint64_t global_unique_id = 1;
 
+struct TEMPLATE_STATE {
+    VkDescriptorUpdateTemplateKHR desc_update_template;
+    safe_VkDescriptorUpdateTemplateCreateInfoKHR *create_info;
```
Comment (chrisforbes) [inline layers/unique_objects.h]: Alternatively, just hold this by value -- the TEMPLATE_STATE is held by unique_ptr already, one layer will do...

File: layers/unique_objects.cpp [hunk 4]
```diff
@@ -643,6 +647,7 @@ VKAPI_ATTR void VKAPI_CALL DestroyDescriptorUpdateTemplateKHR(VkDevice device,
                                                               const VkAllocationCallbacks *pAllocator) {
     layer_data *dev_data = GetLayerDataPtr(get_dispatch_key(device), layer_data_map);
     std::unique_lock<std::mutex> lock(global_lock);
+    dev_data->desc_template_map.erase(descriptorUpdateTemplate);
```
Comment (chrisforbes) [inline layers/unique_objects.cpp]: [Either fix TEMPLATE_STATE so it correctly owns its `create_info` member, or you need to explicitly delete it here.]

File: scripts/helper_file_generator.py [hunk 5]
```diff
@@ -525,9 +525,18 @@ def GenerateSafeStructHelperSource(self):
     # safe_struct source -- create bodies of safe struct helper functions
     def GenerateSafeStructSource(self):
         safe_struct_body = []
+        wsi_structs = ['VkXlibSurfaceCreateInfoKHR',
```
Comment (MarkY-LunarG) [inline scripts/helper_file_generator.py]: This almost seems like it should be a different change.

File: layers/unique_objects.h [hunk 6]
```diff
@@ -55,7 +55,7 @@ struct layer_data {
     VkDebugReportCallbackCreateInfoEXT *tmp_dbg_create_infos;
     VkDebugReportCallbackEXT *tmp_callbacks;
 
-    std::unordered_map<VkDescriptorUpdateTemplateKHR, std::unique_ptr<TEMPLATE_STATE>> desc_template_map;
+    std::unordered_map<uint64_t, std::unique_ptr<TEMPLATE_STATE>> desc_template_map;
```
Comment (mark-lunarg) [inline layers/unique_objects.h]: Done.

File: layers/unique_objects.h (hunk 3 above)
Comment (mark-lunarg) [inline layers/unique_objects.h]: Changed to by-value.

File: layers/unique_objects.cpp [hunk 7]
```diff
@@ -647,25 +649,93 @@ VKAPI_ATTR void VKAPI_CALL DestroyDescriptorUpdateTemplateKHR(VkDevice device,
...
+            if ((create_info->pDescriptorUpdateEntries[i].descriptorType == VK_DESCRIPTOR_TYPE_SAMPLER) ||
+                (create_info->pDescriptorUpdateEntries[i].descriptorType == VK_DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER) ||
+                (create_info->pDescriptorUpdateEntries[i].descriptorType == VK_DESCRIPTOR_TYPE_SAMPLED_IMAGE) ||
+                (create_info->pDescriptorUpdateEntries[i].descriptorType == VK_DESCRIPTOR_TYPE_STORAGE_IMAGE) ||
+                (create_info->pDescriptorUpdateEntries[i].descriptorType == VK_DESCRIPTOR_TYPE_INPUT_ATTACHMENT)) {
+                auto image_entry = reinterpret_cast<VkDescriptorImageInfo *>(entry);
+                size = std::max(size, (char *)offset + sizeof(VkDescriptorImageInfo));
+
+                auto sampler_handle = dev_data->unique_id_mapping[reinterpret_cast<uint64_t &>(image_entry->sampler)];
+                patch_list.emplace_back(std::make_pair(offset, sampler_handle));
```
Comment (mark-lunarg) [inline layers/unique_objects.cpp]: Much nicer!

File: layers/unique_objects.cpp (hunk 1 above)
Comment (mark-lunarg) [inline layers/unique_objects.cpp]: Reworked to let the compiler figure it out -- only copy structures now.

File: layers/unique_objects.cpp (hunk 4 above)
Comment (mark-lunarg) [inline layers/unique_objects.cpp]: Fixed as above

**Comment by mark-lunarg** [comment]
Excellent review feedback incorporated, new version now in place.

File: scripts/helper_file_generator.py (hunk 5 above)
Comment (mark-lunarg) [inline scripts/helper_file_generator.py]: It is -- these changes are in their own commit applied before any of the other changes.

Comment (mark-lunarg) [inline scripts/helper_file_generator.py]: This change is necessary because none of the KHR extensions were getting deep copies for their safe-struct members.  This is required for remapping handles in these extensions.
